Youtube invalid link crash and preview issue

1. If I try to edit `Youtube` message with invalid link the app crashes:

```
E  FATAL EXCEPTION: main
E  Process: io.getstream.chat.sdk, PID: 15733
E  java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.String.equals(java.lang.Object)' on a null object reference
E      at com.getstream.sdk.chat.view.MessageInputView.a(:4)
E      at f.e.a.a.b0.x.a(Unknown Source:4)
E      at androidx.lifecycle.LiveData.a(Unknown Source:29)
E      at androidx.lifecycle.LiveData.b(Unknown Source:44)
E      at androidx.lifecycle.LiveData.b(Unknown Source:14)
E      at d.m.q.b(Unknown Source:0)
E      at androidx.lifecycle.LiveData$a.run(Unknown Source:18)
E      at android.os.Handler.handleCallback(Handler.java:883)
E      at android.os.Handler.dispatchMessage(Handler.java:100)
E      at android.os.Looper.loop(Looper.java:214)
E      at android.app.ActivityThread.main(ActivityThread.java:7356)
E      at java.lang.reflect.Method.invoke(Native Method)
E      at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
E      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

2. `Youtube` and `Unsplash` messages with invalid links looks differently:
<img width="570" alt="Screenshot 2019-10-16 at 16 58 40" src="https://user-images.githubusercontent.com/1069159/66936139-7fd75300-f03d-11e9-86cc-70054d11d656.png">
<img width="514" alt="Screenshot 2019-10-16 at 17 49 35" src="https://user-images.githubusercontent.com/1069159/66936140-7fd75300-f03d-11e9-80aa-d8a4a1698982.png">

Used `Youtube` link: https://www.youtube.com/watch?v=K0x-jSKOoR
Used `Unsplash` link: https://unsplash.com/photos/kAK7A

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,8 @@
 - Fixed read count difference between own and other users' messages
 - Fixed Video length preview
 
+## October 31th, 2019 - 3.2.0
+- Renamed `MessageInputClient` to `MessageInputController` (Adrian)
 ## October 25th, 2019 - 3.2.0
 
 - Added event interceptors to `ChannelListViewModel`

--- a/library/src/main/java/com/getstream/sdk/chat/adapter/AttachmentViewHolder.java
+++ b/library/src/main/java/com/getstream/sdk/chat/adapter/AttachmentViewHolder.java
@@ -64,7 +64,7 @@ public class AttachmentViewHolder extends BaseAttachmentViewHolder {
         boolean hasFile = false;
         boolean hasMedia = false;
         for (Attachment attachment : message.getAttachments()) {
-            if (attachment.getType() == null) continue;
+            if (attachment.getType().equals(ModelType.attach_unknown)) continue;
             if (attachment.getType().equals(ModelType.attach_file)) {
                 hasFile = true;
             } else {
@@ -124,7 +124,7 @@ public class AttachmentViewHolder extends BaseAttachmentViewHolder {
         configAttachViewBackground(lv_attachment_file);
         List<Attachment> attachments = new ArrayList<>();
         for (Attachment attachment : message.getAttachments()) {
-            if (attachment.getType() == null) continue;
+            if (attachment.getType().equals(ModelType.attach_unknown)) continue;
             if (attachment.getType().equals(ModelType.attach_file)) {
                 attachments.add(attachment);
             }
@@ -153,7 +153,7 @@ public class AttachmentViewHolder extends BaseAttachmentViewHolder {
     private void configMediaAttach() {
         List<Attachment> attachments = new ArrayList<>();
         for (Attachment attachment : message.getAttachments()) {
-            if (attachment.getType() == null) continue;
+            if (attachment.getType().equals(ModelType.attach_unknown)) continue;
             if (!attachment.getType().equals(ModelType.attach_file)) {
                 attachments.add(attachment);
             }

--- a/library/src/main/java/com/getstream/sdk/chat/adapter/ChannelListItemViewHolder.java
+++ b/library/src/main/java/com/getstream/sdk/chat/adapter/ChannelListItemViewHolder.java
@@ -165,9 +165,6 @@ public class ChannelListItemViewHolder extends BaseChannelListItemViewHolder {
         }
 
         Attachment attachment = lastMessage.getAttachments().get(0);
-        if (attachment.getType() == null)
-            return;
-
         iv_attachment_type.setVisibility(View.VISIBLE);
 
         String lastMessageText;

--- a/library/src/main/java/com/getstream/sdk/chat/rest/adapter/AttachmentGsonAdapter.java
+++ b/library/src/main/java/com/getstream/sdk/chat/rest/adapter/AttachmentGsonAdapter.java
@@ -1,6 +1,7 @@
 package com.getstream.sdk.chat.rest.adapter;
 
 import com.getstream.sdk.chat.model.Attachment;
+import com.getstream.sdk.chat.model.ModelType;
 import com.getstream.sdk.chat.rest.codecs.GsonConverter;
 import com.google.gson.Gson;
 import com.google.gson.TypeAdapter;
@@ -12,6 +13,9 @@ import java.util.HashMap;
 import java.util.Map;
 
 public class AttachmentGsonAdapter extends TypeAdapter<Attachment> {
+
+    private static final String TAG = AttachmentGsonAdapter.class.getSimpleName();
+
     @Override
     public void write(JsonWriter writer, Attachment attachment) throws IOException {
 
@@ -142,6 +146,10 @@ public class AttachmentGsonAdapter extends TypeAdapter<Attachment> {
             // Set Extra Data
             extraData.put(set.getKey(), set.getValue());
         }
+
+        if (attachment.getType() == null)
+            attachment.setType(ModelType.attach_unknown);
+
         attachment.setExtraData(extraData);
         return attachment;
     }

--- a/library/src/main/java/com/getstream/sdk/chat/utils/MessageInputController.java
+++ b/library/src/main/java/com/getstream/sdk/chat/utils/MessageInputController.java
@@ -38,9 +38,9 @@ import java.io.File;
 import java.util.ArrayList;
 import java.util.List;
 
-public class MessageInputClient {
+public class MessageInputController {
 
-    private static final String TAG = MessageInputClient.class.getSimpleName();
+    private static final String TAG = MessageInputController.class.getSimpleName();
 
     ChannelViewModel viewModel;
     Channel channel;
@@ -58,7 +58,7 @@ public class MessageInputClient {
 
     // region Attachment
 
-    public MessageInputClient(Context context, StreamViewMessageInputBinding binding, ChannelViewModel viewModel, MessageInputStyle style) {
+    public MessageInputController(Context context, StreamViewMessageInputBinding binding, ChannelViewModel viewModel, MessageInputStyle style) {
         this.context = context;
         this.binding = binding;
         this.viewModel = viewModel;

--- a/library/src/main/java/com/getstream/sdk/chat/view/MessageInputView.java
+++ b/library/src/main/java/com/getstream/sdk/chat/view/MessageInputView.java
@@ -257,31 +257,34 @@ public class MessageInputView extends RelativeLayout
 
     // Edit
     private void editMessage(Message message) {
-        if (message == null) {
-            return;
-        }
-
+        if (message == null) return;
+        // Set Text to Inputbox
         binding.etMessage.requestFocus();
         if (!TextUtils.isEmpty(message.getText())) {
             binding.etMessage.setText(message.getText());
             binding.etMessage.setSelection(binding.etMessage.getText().length());
         }
-        if (message.getAttachments() != null && !message.getAttachments().isEmpty()) {
-            for (Attachment attachment : message.getAttachments())
-                attachment.config.setUploaded(true);
 
-            if (message.getAttachments().get(0).getType().equals(ModelType.attach_file)) {
-                String fileType = message.getAttachments().get(0).getMime_type();
-                if (fileType.equals(ModelType.attach_mime_mov) ||
-                        fileType.equals(ModelType.attach_mime_mp4)) {
-                    messageInputClient.onClickOpenSelectMediaView(null, message.getAttachments());
-                } else {
-                    messageInputClient.onClickOpenSelectFileView(null, message.getAttachments());
-                }
+        // Set Attachments to Inputbox
+        if (message.getAttachments() == null
+                || message.getAttachments().isEmpty()
+                || message.getAttachments().get(0).getType().equals(ModelType.attach_giphy)
+                || message.getAttachments().get(0).getType().equals(ModelType.attach_unknown))
+            return;
+
+        for (Attachment attachment : message.getAttachments())
+            attachment.config.setUploaded(true);
+
+        if (message.getAttachments().get(0).getType().equals(ModelType.attach_file)) {
+            String fileType = message.getAttachments().get(0).getMime_type();
+            if (fileType.equals(ModelType.attach_mime_mov) ||
+                    fileType.equals(ModelType.attach_mime_mp4)) {
+                messageInputClient.onClickOpenSelectMediaView(null, message.getAttachments());
             } else {
-                if (!message.getAttachments().get(0).getType().equals(ModelType.attach_giphy))
-                    messageInputClient.onClickOpenSelectMediaView(null, message.getAttachments());
+                messageInputClient.onClickOpenSelectFileView(null, message.getAttachments());
             }
+        } else {
+            messageInputClient.onClickOpenSelectMediaView(null, message.getAttachments());
         }
     }
 

--- a/library/src/main/java/com/getstream/sdk/chat/view/MessageInputView.java
+++ b/library/src/main/java/com/getstream/sdk/chat/view/MessageInputView.java
@@ -266,6 +266,7 @@ public class MessageInputView extends RelativeLayout
         }
 
         // Set Attachments to Inputbox
+
         if (message.getAttachments() == null
                 || message.getAttachments().isEmpty()
                 || message.getAttachments().get(0).getType().equals(ModelType.attach_giphy)
@@ -275,8 +276,9 @@ public class MessageInputView extends RelativeLayout
         for (Attachment attachment : message.getAttachments())
             attachment.config.setUploaded(true);
 
-        if (message.getAttachments().get(0).getType().equals(ModelType.attach_file)) {
-            String fileType = message.getAttachments().get(0).getMime_type();
+        Attachment attachment = message.getAttachments().get(0);
+        if (attachment.getType().equals(ModelType.attach_file)) {
+            String fileType = attachment.getMime_type();
             if (fileType.equals(ModelType.attach_mime_mov) ||
                     fileType.equals(ModelType.attach_mime_mp4)) {
                 messageInputClient.onClickOpenSelectMediaView(null, message.getAttachments());

--- a/library/src/main/java/com/getstream/sdk/chat/view/MessageInputView.java
+++ b/library/src/main/java/com/getstream/sdk/chat/view/MessageInputView.java
@@ -44,7 +44,7 @@ import com.getstream.sdk.chat.rest.interfaces.MessageCallback;
 import com.getstream.sdk.chat.rest.response.MessageResponse;
 import com.getstream.sdk.chat.utils.Constant;
 import com.getstream.sdk.chat.utils.GridSpacingItemDecoration;
-import com.getstream.sdk.chat.utils.MessageInputClient;
+import com.getstream.sdk.chat.utils.MessageInputController;
 import com.getstream.sdk.chat.utils.Utils;
 import com.getstream.sdk.chat.viewmodel.ChannelViewModel;
 
@@ -95,8 +95,8 @@ public class MessageInputView extends RelativeLayout
      * The viewModel for handling typing etc.
      */
     private ChannelViewModel viewModel;
-    // TODO Rename, totally not clear what this does
-    private MessageInputClient messageInputClient;
+    
+    private MessageInputController messageInputController;
 
     public MessageInputView(Context context) {
         super(context);
@@ -172,7 +172,7 @@ public class MessageInputView extends RelativeLayout
             attachment.setTitleLink(url);
             attachment.setTitle(inputContentInfo.getDescription().getLabel().toString());
             attachment.setType(ModelType.attach_giphy);
-            messageInputClient.setSelectedAttachments(Arrays.asList(attachment));
+            messageInputController.setSelectedAttachments(Arrays.asList(attachment));
             onSendMessage("", viewModel.isEditing());
             return true;
         });
@@ -199,7 +199,7 @@ public class MessageInputView extends RelativeLayout
                     return true;
                 }
                 if (viewModel.isEditing()) {
-                    messageInputClient.onClickCloseBackGroundView();
+                    messageInputController.onClickCloseBackGroundView();
                     initSendMessage();
                     return true;
                 }
@@ -209,7 +209,7 @@ public class MessageInputView extends RelativeLayout
                 }
 
                 if (binding.clTitle.getVisibility() == VISIBLE) {
-                    messageInputClient.onClickCloseBackGroundView();
+                    messageInputController.onClickCloseBackGroundView();
                     initSendMessage();
                     return true;
                 }
@@ -237,7 +237,7 @@ public class MessageInputView extends RelativeLayout
                     binding.llComposer.setBackground(style.getInputEditBackground());
                     binding.ivOpenAttach.setImageDrawable(style.getAttachmentButtonIcon(true));
                     binding.ivSend.setImageDrawable(style.getInputButtonIcon(true));
-                    messageInputClient.onClickOpenBackGroundView(MessageInputType.EDIT_MESSAGE);
+                    messageInputController.onClickOpenBackGroundView(MessageInputType.EDIT_MESSAGE);
                     break;
             }
         });
@@ -281,18 +281,18 @@ public class MessageInputView extends RelativeLayout
             String fileType = attachment.getMime_type();
             if (fileType.equals(ModelType.attach_mime_mov) ||
                     fileType.equals(ModelType.attach_mime_mp4)) {
-                messageInputClient.onClickOpenSelectMediaView(null, message.getAttachments());
+                messageInputController.onClickOpenSelectMediaView(null, message.getAttachments());
             } else {
-                messageInputClient.onClickOpenSelectFileView(null, message.getAttachments());
+                messageInputController.onClickOpenSelectFileView(null, message.getAttachments());
             }
         } else {
-            messageInputClient.onClickOpenSelectMediaView(null, message.getAttachments());
+            messageInputController.onClickOpenSelectMediaView(null, message.getAttachments());
         }
     }
 
     private void initAttachmentUI() {
         // TODO: make the attachment UI into it's own view and allow you to change it.
-        messageInputClient = new MessageInputClient(getContext(), binding, this.viewModel, style);
+        messageInputController = new MessageInputController(getContext(), binding, this.viewModel, style);
         binding.rvMedia.setLayoutManager(new GridLayoutManager(getContext(), 4, RecyclerView.VERTICAL, false));
         binding.rvMedia.hasFixedSize();
         binding.rvComposer.setLayoutManager(new GridLayoutManager(getContext(), 1, LinearLayoutManager.HORIZONTAL, false));
@@ -301,18 +301,18 @@ public class MessageInputView extends RelativeLayout
         boolean includeEdge = false;
         binding.rvMedia.addItemDecoration(new GridSpacingItemDecoration(spanCount, spacing, includeEdge));
         binding.tvClose.setOnClickListener(v -> {
-            messageInputClient.onClickCloseBackGroundView();
+            messageInputController.onClickCloseBackGroundView();
             if (viewModel.isEditing()) {
                 initSendMessage();
                 clearFocus();
             }
         });
 
-        binding.llMedia.setOnClickListener(v -> messageInputClient.onClickOpenSelectMediaView(v, null));
+        binding.llMedia.setOnClickListener(v -> messageInputController.onClickOpenSelectMediaView(v, null));
 
         binding.llCamera.setOnClickListener(v -> {
             Utils.setButtonDelayEnable(v);
-            messageInputClient.onClickCloseBackGroundView();
+            messageInputController.onClickCloseBackGroundView();
 
             Intent takePictureIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
             ContentValues values = new ContentValues();
@@ -329,7 +329,7 @@ public class MessageInputView extends RelativeLayout
             if (this.openCameraViewListener != null)
                 openCameraViewListener.openCameraView(chooserIntent, Constant.CAPTURE_IMAGE_REQUEST_CODE);
         });
-        binding.llFile.setOnClickListener(v -> messageInputClient.onClickOpenSelectFileView(v, null));
+        binding.llFile.setOnClickListener(v -> messageInputController.onClickOpenSelectFileView(v, null));
     }
 
     // endregion
@@ -345,7 +345,7 @@ public class MessageInputView extends RelativeLayout
                     e.printStackTrace();
                 }
                 if (file != null) {
-                    messageInputClient.progressCapturedMedia(getContext(), imageUri, true);
+                    messageInputController.progressCapturedMedia(getContext(), imageUri, true);
                     imageUri = null;
                 }
                 return;
@@ -354,10 +354,10 @@ public class MessageInputView extends RelativeLayout
                 Uri uri = data.getData();
                 if (uri == null) {
                     if (imageUri != null)
-                        messageInputClient.progressCapturedMedia(getContext(), imageUri, true);
+                        messageInputController.progressCapturedMedia(getContext(), imageUri, true);
                     imageUri = null;
                 } else {
-                    messageInputClient.progressCapturedMedia(getContext(), uri, false);
+                    messageInputController.progressCapturedMedia(getContext(), uri, false);
                 }
             } catch (Exception e) {
                 e.printStackTrace();
@@ -381,7 +381,7 @@ public class MessageInputView extends RelativeLayout
 
             style.setShowAttachmentButton(granted);
             if (granted)
-                messageInputClient.onClickOpenBackGroundView(MessageInputType.ADD_FILE);
+                messageInputController.onClickOpenBackGroundView(MessageInputType.ADD_FILE);
             else
                 binding.ivOpenAttach.setVisibility(GONE);
 
@@ -432,7 +432,7 @@ public class MessageInputView extends RelativeLayout
         } else if (id == R.id.iv_openAttach) {
             binding.setIsAttachFile(true);
             if (style.isPermissionSet() || isGrantedPermissions())
-                messageInputClient.onClickOpenBackGroundView(MessageInputType.ADD_FILE);
+                messageInputController.onClickOpenBackGroundView(MessageInputType.ADD_FILE);
             else if(permissionRequestListener != null)
                 permissionRequestListener.openPermissionRequest();
         }
@@ -456,7 +456,7 @@ public class MessageInputView extends RelativeLayout
             viewModel.keystroke();
         }
         // detect commands
-        messageInputClient.checkCommand(messageText);
+        messageInputController.checkCommand(messageText);
         String s_ = messageText.replaceAll("\\s+","");
         if (TextUtils.isEmpty(s_))
             binding.setActiveMessageSend(false);
@@ -483,7 +483,7 @@ public class MessageInputView extends RelativeLayout
         Message m = new Message();
         m.setStatus(null);
         m.setText(input);
-        m.setAttachments(messageInputClient.getSelectedAttachments());
+        m.setAttachments(messageInputController.getSelectedAttachments());
         // set the thread id if we are viewing a thread
         if (viewModel.isThread())
             m.setParentId(viewModel.getThreadParentMessage().getValue().getId());
@@ -498,7 +498,7 @@ public class MessageInputView extends RelativeLayout
         if (isEdit) {
             Message message = getEditMessage();
             message.setText(input);
-            List<Attachment>newAttachments = messageInputClient.getSelectedAttachments();
+            List<Attachment>newAttachments = messageInputController.getSelectedAttachments();
             if (newAttachments != null
                     && !newAttachments.isEmpty()){
                 List<Attachment>attachments = message.getAttachments();
@@ -551,7 +551,7 @@ public class MessageInputView extends RelativeLayout
     }
 
     private void initSendMessage() {
-        messageInputClient.initSendMessage();
+        messageInputController.initSendMessage();
         viewModel.setEditMessage(null);
         binding.etMessage.setText("");
     }

--- a/library/src/main/java/com/getstream/sdk/chat/view/MessageListView.java
+++ b/library/src/main/java/com/getstream/sdk/chat/view/MessageListView.java
@@ -181,7 +181,7 @@ public class MessageListView extends RecyclerView {
             @Override
             public Drawable getDrawableForAttachment(Message message, Boolean mine, List<MessageViewHolderFactory.Position> positions, Attachment attachment) {
                 if (attachment == null
-                        || attachment.getType() == null
+                        || attachment.getType().equals(ModelType.attach_unknown)
                         || attachment.getType().equals(ModelType.attach_file)) 
                     return null;
 
@@ -577,8 +577,7 @@ public class MessageListView extends RecyclerView {
     public void showAttachment(Message message, Attachment attachment) {
         String url = null;
         String type = null;
-        if (attachment == null
-                || attachment.getType() == null)
+        if (attachment == null)
             return;
 
         switch (attachment.getType()) {
@@ -619,6 +618,11 @@ public class MessageListView extends RecyclerView {
                 url = attachment.getUrl();
                 break;
         }
+        if (TextUtils.isEmpty(url)){
+            Utils.showMessage(getContext(), getContext().getString(R.string.stream_attachment_invalid_url));
+            return;
+        }
+
         if (type == null) type = attachment.getType();
         Intent intent = new Intent(getContext(), AttachmentActivity.class);
         intent.putExtra("type", type);

--- a/library/src/main/res/values/strings.xml
+++ b/library/src/main/res/values/strings.xml
@@ -39,5 +39,5 @@
     <string name="stream_last_message_attachment_giphy">/giphy</string>
     <!--Attachment View-->
     <string name="stream_attachment_load_failed_unknown">The load failed due to an unknown error.</string>
-
+    <string name="stream_attachment_invalid_url">The load failed due to the invalid url.</string>
 </resources>

